### PR TITLE
feat: support to trace the process of sync stream

### PR DESCRIPTION
**Pull request by Anyc66666666**

### CODE REVIEW

File: src/greptimeai/collector/__init__.py [hunk 1]
```diff
@@ -81,3 +81,15 @@ def collect_metrics(self, span_attrs: Dict[str, Any], attrs: Optional[Attributes
             completion_cost=completion_cost,
             attrs=attrs,
         )
+
+    def add_span_event(
```
Comment (yuanbohan) [inline src/greptimeai/collector/__init__.py:85]: if this method has the same signature of `_collector.add_span_event`, we can just call it without a new method.

If a new method will make method signature cleaner, maybe we can give a new one

File: src/greptimeai/extractor/openai_extractor/__init__.py [hunk 2]
```diff
@@ -80,7 +80,7 @@ def get_user_id(**kwargs) -> Optional[str]:
 
     @staticmethod
     def extract_usage(
-        model: Optional[str], usage: Optional[Dict[str, int]]
+        model: Optional[str], usage: Any
```
Comment (yuanbohan) [inline src/greptimeai/extractor/openai_extractor/__init__.py:83]: `usage: Any` may make static type checking weak to find potential error. Is it ok to keep the original `Optional[Dict[str, int]]`?

File: src/greptimeai/utils/openai/token.py [hunk 3]
```diff
@@ -216,3 +218,21 @@ def get_openai_audio_cost_for_whisper(seconds: int) -> float:
     """
     cost = 0.006 * (seconds / 60)
     return round(cost, 6)
+
+def count_tokens(model, text):
+    encoding = None
+    try:
+        tiktoken = importlib.import_module("tiktoken")
```
Comment (yuanbohan) [inline src/greptimeai/utils/openai/token.py:225]: `tiktoken` is the necessary dependency so far, just use it directly

Comment (yuanbohan) [reply src/greptimeai/utils/openai/token.py:225]: and greptimeai already has a similar function, `num_tokens_from_messages`?

File: src/greptimeai/patcher/openai_patcher/base.py [hunk 4]
```diff
@@ -161,82 +182,180 @@ def patch(self):
...
+            if not prompt_cost_span:
+                extraction.span_attributes[_PROMPT_COST_LABEl] = self.prompt_cost
+            if usage and _PROMPT_COST_LABEl not in usage:
+                extraction.event_attributes["usage"][
+                    _PROMPT_COST_LABEl
+                ] = self.prompt_cost
+
+        return extraction
+
+    def _trace_stream(self, span_id, span_name, start, resp, ex):
```
Comment (yuanbohan) [inline src/greptimeai/patcher/openai_patcher/base.py:210]: the logic is a little bit too complicated for maintainance

File: src/greptimeai/utils/openai/stream.py [hunk 5]
```diff
@@ -0,0 +1,11 @@
+from openai._streaming import Stream, AsyncStream
+
+
+class StreamUtil:
+    @staticmethod
+    def is_stream(obj):
+        return obj and isinstance(obj, Stream)
+
+    @staticmethod
+    def is_async_stream(obj):
```
[the remark this answers is not in the document]
Comment (Anyc66666666) [reply src/greptimeai/utils/openai/stream.py:10]: in async wrapper func

**Pull request closed without merging**